### PR TITLE
Repetitive calls - get events for multiple event types in one getLogs call

### LINKED ISSUE #3443: Repetitive RPC calls by EventScanner
From Alchemy stats of a node running v7.2.0-dev.11, filtering for `eth_getLogs` exclusively:
image

Observations:
* 24 calls each 5 minutes (1 each 12.5 seconds). For prior version, 8640 calls to eth_getLogs during last 24h (1 call each 10 seconds). So it seems there's not much improvement yet.
* For some reason, there are several simultaneous calls scanning the same range, normally 3 times. For example, look at the block ranges in the first 6 `eth_getLogs` calls: the first 3 have the same range, the next 3 the same, etc. I say simultaneous because, at least according to Alchemy, these calls happen in blocks of 6 that happen roughly at the same time (at least, within a second). The next block of 6 happened one minute before, etc.

### PR DESCRIPTION
**Pull request by derekpierre**
**Type of PR:**
- [ ] Bugfix
- [ ] Feature
- [ ] Documentation
- [X] Other

**Required reviews:** 
- [ ] 1
- [ ] 2
- [X] 3

**What this does:**
Closes #3443 

Previously we were call eth_getLogs individually for each event type: StartRitual, StartAggregationRound, EndRitual - https://github.com/nucypher/nucypher/blob/v7.2.x/nucypher/utilities/events.py#L239 . 

Instead we can make one eth_getLogs call for all 3 event types. Our volume of rpc calls should now be 1/3 of what it was previously.

**Issues fixed/closed:**
> - Fixes #...

**Why it's needed:**
> Explain how this PR fits in the greater context of the NuCypher Network.
> E.g., if this PR address a `nucypher/productdev` issue, let reviewers know!

**Notes for reviewers:**
> What should reviewers focus on? 
> Is there a particular commit/function/section of your PR that requires more attention from reviewers?

### CODE REVIEW

File: nucypher/utilities/events.py [hunk 1]
```diff
@@ -378,45 +374,54 @@ def _fetch_events_for_all_contracts(
...
+    event_topics_to_abis = {}
+    for event_type in events:
+        event_abi = event_type._get_event_abi()
+        event_topic = encode_hex(event_abi_to_log_topic(event_abi))  # type: ignore
+        topics.add(event_topic)
+        event_topics_to_abis[event_topic] = event_abi
+
+    event_filter_params = {
+        "address": contract.address,
+        "topics": [list(topics)],
```
Comment (cygnusv) [inline nucypher/utilities/events.py:394]:
```suggestion
        "topics": list(topics),
```
This should be a list of topics, right?

File: nucypher/utilities/events.py [hunk 2]
```diff
@@ -378,45 +374,54 @@ def _fetch_events_for_all_contracts(
...
+        "fromBlock": from_block,
+    }
+    if to_block is not None:
+        event_filter_params["toBlock"] = to_block
+
+    logger.debug(
+        f"Querying eth_get_logs with the following parameters: {event_filter_params}"
```
Comment (cygnusv) [inline nucypher/utilities/events.py:401]:
```suggestion
        f"Querying eth_getLogs with the following parameters: {event_filter_params}"
```

File: nucypher/utilities/events.py [hunk 3]
```diff
@@ -361,12 +360,9 @@ def scan(
...
-def _fetch_events_for_all_contracts(
-        web3,
-        event,
-        argument_filters: dict,
-        from_block: int,
-        to_block: int) -> Iterable:
+def _fetch_events_for_contract(
+    web3, contract, events, from_block: int, to_block: int
+) -> Iterable:
     """Get events using eth_get_logs API.
```
Comment (cygnusv) [inline nucypher/utilities/events.py:366]:
```suggestion
    """Get events using eth_getLogs API.
```

**Review by manumonti — APPROVED** [review]
Good improvement! LGTM 👌

**Comment by KPrasch** [comment]
Looking mostly good! This issues appears to be strictly an optimization, did you mean to base it over v7.2.x (frozen)?

File: nucypher/utilities/events.py (hunk 1 above)
Comment (derekpierre) [reply nucypher/utilities/events.py:394]: No. It is a list of list on purpose. See https://web3py.readthedocs.io/en/stable/filters.html#event-log-filters.

For searching for events using topics the first element is the set of topics, and subsequent entries are arguments to filter the events by.

In our case we want multiple topics (events) to be searched for at the same time with no arguments (essentially don't filter on any arguments) so:
`[[A, B, C]]` means search for events with topics A OR B OR C

`[A, B, C]` would mean search for events with topic A, and arguments B, C.

**Comment by derekpierre** [comment]
> Looking mostly good! This issues appears to be strictly an optimization, did you mean to base it over v7.2.x (frozen)?

Yes, I did mean to base it over `v7.2.x` for a couple reasons:
1. Of all of the RPC call reduction work we have done on v7.2.x thus far, this will have the biggest impact. It should reduce the number of rpc calls to 1/3 of what it was before which is very significant, and was one of the goals of 7.2.
2. It is a very specific change in a limited area (single file) of the code base.

**Comment by derekpierre** [comment]
Updated screenshot of alchemy endpoint being used by lynx-3 which is currently running the fix.
Screenshot 2024-02-13 at 8 19 18 AM

**Review by KPrasch — APPROVED** [review]

**Pull request merged**